sky/serve/load_balancer.py

```diff
@@ -223,10 +322,250 @@ async def startup():
+            'filter_chains': [{
+                'filters': filters
+            }]
+        }
+
+        # These endpoints are the backend IP addresses that the cluster defined
+        # below can load balance over.
+        lb_endpoints = []
+        for url in replicas:
+            # TODO(ejj) when Envoy is the only load balancer choice, we should
+            # just have the controller send us the IP and port rather than
+            # bundling them together in a url.
+            parsed_url = urlparse(url)
+            lb_endpoints.append({
+                'endpoint': {
+                    'address': {
+                        'socket_address': {
+                            'address': parsed_url.hostname,
+                            'port_value': parsed_url.port,
+                        }
+                    }
+                }
+            })
+
+        # A cluster is usually a group of endpoints that can be load balanced
+        # over.  This cluster contains the replicas of the service, and is
+        # handed traffic by the above listener.
+        cluster = {
+            'name': 'cluster',
+            'connect_timeout': '0.25s',
+            'type': 'STATIC',
+            'lb_policy': 'ROUND_ROBIN',  # TODO(ejj) weighted least request
+            'load_assignment': {
+                'cluster_name': 'cluster',
+                'endpoints': [{
+                    'lb_endpoints': lb_endpoints
+                }]
+            },
+        }
+
+        config = {
+            'static_resources': {
+                'listeners': [listener],
+                'clusters': [cluster],
+            }
+        }
+
+        return yaml.dump(config, default_flow_style=False)
+
+    async def run(self):
+        await self.update_envoy_config([])
+
+        # Because docker can take some time to come up, we make multiple
+        # attempts to start before giving up.
+        envoy_started = False
+        logger.info('Starting envoy %s', self.envoy_name)
+        for _ in range(30):
+            await asyncio.sleep(5)
+            envoy_started = await self._start_envoy()
+            if envoy_started:
+                break
+
+        if not envoy_started:
+            error = f'Failed to start envoy {self.envoy_name}'
+            logger.error(error)
+            raise RuntimeError(error)
+
+        while True:
+            await asyncio.sleep(constants.LB_CONTROLLER_SYNC_INTERVAL_SECONDS)
+
+            ready_replica_urls = await self._controller_sync({})
+            if ready_replica_urls is None:
+                continue
+
+            if not await self.update_envoy_config(ready_replica_urls):
+                continue
+
+            proc = await asyncio.create_subprocess_exec('docker', 'restart',
+                                                        self.envoy_name)
+            return_code = await proc.wait()
+            if return_code != 0:
+                logger.warning('failed to restart envoy %s', self.envoy_name)
```

Comment:
`docker restart` is instant.  It just reboots the process, it doesn't tear down and recreate the container.

The main cost here, is any in-flight connections are dropped.   As @Michaelvll  mentioned for AI workloads this may not be acceptable.

IMO it's instant enough that it's probably good enough to merge as the non-default option and have hot-restart fast follow, but not sure.

Reply:
To be clear, I didn't formally measure this.  When I say instant I mean like < 1s but that's just a gut feel from trying it manually

Reply:
latest version of the PR gets rid of the restart.